In [1]:
import pandas as pd
import numpy as np
import plotly.express as px
import warnings
warnings.filterwarnings('ignore')

In [2]:
df=pd.read_csv("/content/Australian_Vehicle_Prices_Data_Cleaned.csv")

In [3]:
df

Brand    Year    Model               Car/Suv  \
0          Ssangyong  2022.0   Rexton  Sutherland Isuzu Ute   
1                 MG  2022.0      MG3             Hatchback   
2                BMW  2022.0     430I                 Coupe   
3      Mercedes-Benz  2011.0     E500                 Coupe   
4            Renault  2022.0   Arkana                   SUV   
...              ...     ...      ...                   ...   
14582        Hyundai  2015.0      I30                 Wagon   
14583         Nissan  2021.0  Qashqai                   SUV   
14584         Suzuki  2022.0   Vitara   Bremer Suzuki - New   
14585  Mercedes-Benz  2016.0   GLC250                   SUV   
14586  Mercedes-Benz  2021.0     C200                 Sedan   

                                           Title UsedOrNew Transmission  \
0           2022 Ssangyong Rexton Ultimate (awd)      DEMO    Automatic   
1      2022 MG MG3 Auto Excite (with Navigation)      USED    Automatic   
2                          2022 BMW 430I M Sport      USED    Automatic   
3               2011 Mercedes-Benz E500 Elegance      USED    Automatic   
4                     2022 Renault Arkana Intens      USED    Automatic   
...                                          ...       ...          ...   
14582     2015 Hyundai I30 Tourer Active 1.6 GDI      USED    Automatic   
14583                     2021 Nissan Qashqai TI      USED    Automatic   
14584                    2022 Suzuki Vitara 1.6L      DEMO    Automatic   
14585                  2016 Mercedes-Benz GLC250      USED    Automatic   
14586                    2021 Mercedes-Benz C200      USED    Automatic   

             Engine DriveType  FuelType  ...  ColourExtInt  \
0      4 cyl, 2.2 L       AWD    Diesel  ...         White   
1      4 cyl, 1.5 L     Front   Premium  ...         Black   
2        4 cyl, 2 L      Rear   Premium  ...          Grey   
3      8 cyl, 5.5 L      Rear   Premium  ...         White   
4      4 cyl, 1.3 L     Front  Unleaded  ...          Grey   
...             ...       ...       ...  ...           ...   
14582  4 cyl, 1.6 L     Front  Unleaded  ...        Silver   
14583    4 cyl, 2 L     Front  Unleaded  ...         White   
14584  4 cyl, 1.6 L     Front  Unleaded  ...         Other   
14585    4 cyl, 2 L       AWD   Premium  ...          Grey   
14586    4 cyl, 2 L      Rear  Unleaded  ...         Other   

                  Location CylindersinEngine   BodyType  Doors Seats  \
0           Caringbah, NSW               4.0        SUV    4.0   7.0   
1           Brookvale, NSW               4.0  Hatchback    5.0   5.0   
2            Sylvania, NSW               4.0      Coupe    2.0   4.0   
3        Mount Druitt, NSW               8.0      Coupe    2.0   4.0   
4         Castle Hill, NSW               4.0        SUV    4.0   5.0   
...                    ...               ...        ...    ...   ...   
14582  Oakleigh South, VIC               4.0      Wagon    4.0   5.0   
14583         Rocklea, QLD               4.0        SUV    4.0   5.0   
14584         Dinmore, QLD               4.0        SUV    4.0   5.0   
14585          Albion, QLD               4.0        SUV    4.0   5.0   
14586          Yatala, QLD               4.0      Sedan    4.0   5.0   

          Price         Brand & Model  LitersinEngine State  
0       51990.0      Ssangyong Rexton             2.2   NSW  
1       19990.0                MG MG3             1.5   NSW  
2      108988.0              BMW 430I             2.0   NSW  
3       32990.0    Mercedes-Benz E500             5.5   NSW  
4       34990.0        Renault Arkana             1.3   NSW  
...         ...                   ...             ...   ...  
14582   21450.0           Hyundai I30             1.6   VIC  
14583   36770.0        Nissan Qashqai             2.0   QLD  
14584   35280.0         Suzuki Vitara             1.6   QLD  
14585   41888.0  Mercedes-Benz GLC250             2.0   QLD  
14586   65888.0    Mercedes-Benz C200             2.0   QLD  

[1

In [4]:
df.columns

Index(['Brand', 'Year', 'Model', 'Car/Suv', 'Title', 'UsedOrNew',
       'Transmission', 'Engine', 'DriveType', 'FuelType', 'FuelConsumption',
       'Kilometres', 'ColourExtInt', 'Location', 'CylindersinEngine',
       'BodyType', 'Doors', 'Seats', 'Price', 'Brand & Model',
       'LitersinEngine', 'State'],
      dtype='object')

In [5]:
df.drop(['Brand','Model','Car/Suv','Title','Engine','Location'],axis=1,inplace=True)

In [6]:
df

Year UsedOrNew Transmission DriveType  FuelType  FuelConsumption  \
0      2022.0      DEMO    Automatic       AWD    Diesel              8.7   
1      2022.0      USED    Automatic     Front   Premium              6.7   
2      2022.0      USED    Automatic      Rear   Premium              6.6   
3      2011.0      USED    Automatic      Rear   Premium             11.0   
4      2022.0      USED    Automatic     Front  Unleaded              6.0   
...       ...       ...          ...       ...       ...              ...   
14582  2015.0      USED    Automatic     Front  Unleaded              6.9   
14583  2021.0      USED    Automatic     Front  Unleaded              6.9   
14584  2022.0      DEMO    Automatic     Front  Unleaded              6.2   
14585  2016.0      USED    Automatic       AWD   Premium              7.2   
14586  2021.0      USED    Automatic      Rear  Unleaded              7.0   

       Kilometres ColourExtInt  CylindersinEngine   BodyType  Doors  Seats  \
0          5595.0        White                4.0        SUV    4.0    7.0   
1            16.0        Black                4.0  Hatchback    5.0    5.0   
2          8472.0         Grey                4.0      Coupe    2.0    4.0   
3        136517.0        White                8.0      Coupe    2.0    4.0   
4          1035.0         Grey                4.0        SUV    4.0    5.0   
...           ...          ...                ...        ...    ...    ...   
14582     54359.0       Silver                4.0      Wagon    4.0    5.0   
14583     32977.0        White                4.0        SUV    4.0    5.0   
14584      5165.0        Other                4.0        SUV    4.0    5.0   
14585     85525.0         Grey                4.0        SUV    4.0    5.0   
14586     31852.0        Other                4.0      Sedan    4.0    5.0   

          Price         Brand & Model  LitersinEngine State  
0       51990.0      Ssangyong Rexton             2.2   NSW  
1       19990.0                MG MG3             1.5   NSW  
2      108988.0              BMW 430I             2.0   NSW  
3       32990.0    Mercedes-Benz E500             5.5   NSW  
4       34990.0        Renault Arkana             1.3   NSW  
...         ...                   ...             ...   ...  
14582   21450.0           Hyundai I30             1.6   VIC  
14583   36770.0        Nissan Qashqai             2.0   QLD  
14584   35280.0         Suzuki Vitara             1.6   QLD  
14585   41888.0  Mercedes-Benz GLC250             2.0   QLD  
14586   65888.0    Mercedes-Benz C200             2.0   QLD  

[14587 rows x 16 columns]

In [7]:
!pip install category_encoders

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 81.9/81.9 kB 1.7 MB/s eta 0:00:00


In [8]:
from sklearn.preprocessing import FunctionTransformer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline , make_pipeline

from sklearn.preprocessing import StandardScaler,MinMaxScaler,RobustScaler
from sklearn.preprocessing import OneHotEncoder
import category_encoders as ce

from sklearn.linear_model import LinearRegression,Lasso,Ridge
from sklearn.svm import SVR
from sklearn.preprocessing import PolynomialFeatures
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import BaggingRegressor
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor

from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error

from sklearn.decomposition import PCA
from sklearn.feature_selection import SequentialFeatureSelector,SelectFromModel

from sklearn.model_selection import cross_validate,cross_val_score
from sklearn.model_selection import GridSearchCV,RandomizedSearchCV



In [9]:
for col in df.select_dtypes('object').columns:
  print(col,' ',df[col].nunique())

UsedOrNew   3
Transmission   2
DriveType   5
FuelType   7
ColourExtInt   7
BodyType   10
Brand & Model   657
State   9


In [10]:
x=df.drop(['Price'],axis=1)
y=df['Price']

In [11]:
x

Year UsedOrNew Transmission DriveType  FuelType  FuelConsumption  \
0      2022.0      DEMO    Automatic       AWD    Diesel              8.7   
1      2022.0      USED    Automatic     Front   Premium              6.7   
2      2022.0      USED    Automatic      Rear   Premium              6.6   
3      2011.0      USED    Automatic      Rear   Premium             11.0   
4      2022.0      USED    Automatic     Front  Unleaded              6.0   
...       ...       ...          ...       ...       ...              ...   
14582  2015.0      USED    Automatic     Front  Unleaded              6.9   
14583  2021.0      USED    Automatic     Front  Unleaded              6.9   
14584  2022.0      DEMO    Automatic     Front  Unleaded              6.2   
14585  2016.0      USED    Automatic       AWD   Premium              7.2   
14586  2021.0      USED    Automatic      Rear  Unleaded              7.0   

       Kilometres ColourExtInt  CylindersinEngine   BodyType  Doors  Seats  \
0          5595.0        White                4.0        SUV    4.0    7.0   
1            16.0        Black                4.0  Hatchback    5.0    5.0   
2          8472.0         Grey                4.0      Coupe    2.0    4.0   
3        136517.0        White                8.0      Coupe    2.0    4.0   
4          1035.0         Grey                4.0        SUV    4.0    5.0   
...           ...          ...                ...        ...    ...    ...   
14582     54359.0       Silver                4.0      Wagon    4.0    5.0   
14583     32977.0        White                4.0        SUV    4.0    5.0   
14584      5165.0        Other                4.0        SUV    4.0    5.0   
14585     85525.0         Grey                4.0        SUV    4.0    5.0   
14586     31852.0        Other                4.0      Sedan    4.0    5.0   

              Brand & Model  LitersinEngine State  
0          Ssangyong Rexton             2.2   NSW  
1                    MG MG3             1.5   NSW  
2                  BMW 430I             2.0   NSW  
3        Mercedes-Benz E500             5.5   NSW  
4            Renault Arkana             1.3   NSW  
...                     ...             ...   ...  
14582           Hyundai I30             1.6   VIC  
14583        Nissan Qashqai             2.0   QLD  
14584         Suzuki Vitara             1.6   QLD  
14585  Mercedes-Benz GLC250             2.0   QLD  
14586    Mercedes-Benz C200             2.0   QLD  

[14587 rows x 15 columns]

In [12]:
y

0         51990.0
1         19990.0
2        108988.0
3         32990.0
4         34990.0
           ...   
14582     21450.0
14583     36770.0
14584     35280.0
14585     41888.0
14586     65888.0
Name: Price, Length: 14587, dtype: float64

In [13]:
px.histogram(np.log(y))


In [14]:
log_y=np.log(y)

In [15]:
px.histogram(log_y)

In [16]:
log_col=['Kilometres']
ohe_col=['UsedOrNew','Transmission','DriveType']
be_col=['FuelType','ColourExtInt','BodyType','Brand & Model','State']

In [ ]:
# log_fun= FunctionTransformer(np.log1p)

In [17]:
transform=ColumnTransformer(transformers=[
    ('log',FunctionTransformer(np.log1p),log_col),
    ('Ohe',OneHotEncoder(sparse_output=False,drop='first'),ohe_col),
    ('Be',ce.BinaryEncoder(),be_col)
],remainder='passthrough')

In [18]:
transform

ColumnTransformer(remainder='passthrough',
                  transformers=[('log',
                                 FunctionTransformer(func=<ufunc 'log1p'>),
                                 ['Kilometres']),
                                ('Ohe',
                                 OneHotEncoder(drop='first',
                                               sparse_output=False),
                                 ['UsedOrNew', 'Transmission', 'DriveType']),
                                ('Be', BinaryEncoder(),
                                 ['FuelType', 'ColourExtInt', 'BodyType',
                                  'Brand & Model', 'State'])])

In [19]:
models=[
    ('lr',LinearRegression()),
    ('Lasso',Lasso()),
    ('Ridge',Ridge()),
    ('SVC',SVR()),
    ('Knn',KNeighborsRegressor()),
    ('DT',DecisionTreeRegressor()),
    ('Bagging',BaggingRegressor()),
    ('RT',RandomForestRegressor()),
    ('XGB',XGBRegressor())
]


**compare the result from Y and Log(Y)**

In [20]:
for model in models:
  print(model[0])
  PL=make_pipeline(transform,
                   StandardScaler(),

                   model[1])

  cv=cross_validate(estimator=PL,X=x,y=y,cv=7,return_train_score=True,scoring=['r2','neg_mean_squared_error'])
  print(model[0],'train r2 score = ',cv['train_r2'].mean())
  print(model[0],'test r2 score = ',cv['test_r2'].mean())
  print(model[0],'train mean_squared_error = ',cv['train_neg_mean_squared_error'].mean())
  print(model[0],'test mean_squared_error = ',cv['test_neg_mean_squared_error'].mean())
  print('-'*30)

lr
lr train r2 score =  0.5068480843948319
lr test r2 score =  0.47703628691542604
lr train mean_squared_error =  -414936419.2363977
lr test mean_squared_error =  -445335367.41919196
------------------------------
Lasso
Lasso train r2 score =  0.5068479979712667
Lasso test r2 score =  0.4770716257609794
Lasso train mean_squared_error =  -414936491.6815547
Lasso test mean_squared_error =  -445311691.6251921
------------------------------
Ridge
Ridge train r2 score =  0.5068480772863657
Ridge test r2 score =  0.4770414060278484
Ridge train mean_squared_error =  -414936425.209825
Ridge test mean_squared_error =  -445333040.61819786
------------------------------
SVC
SVC train r2 score =  -0.035780377191896404
SVC test r2 score =  -0.04138917850681164
SVC train mean_squared_error =  -870589588.6329815
SVC test mean_squared_error =  -872456865.5778592
------------------------------
Knn
Knn train r2 score =  0.7765042630213832
Knn test r2 score =  0.6208247133008212
Knn train mean_squared_er

In [21]:
for model in models:
  print(model[0])
  PL=make_pipeline(transform,
                   StandardScaler(),

                   model[1])

  cv=cross_validate(estimator=PL,X=x,y=log_y,cv=7,return_train_score=True,scoring=['r2','neg_mean_squared_error'])
  print(model[0],'train r2 score = ',cv['train_r2'].mean())
  print(model[0],'test r2 score = ',cv['test_r2'].mean())
  print(model[0],'train mean_squared_error = ',cv['train_neg_mean_squared_error'].mean())
  print(model[0],'test mean_squared_error = ',cv['test_neg_mean_squared_error'].mean())
  print('-'*30)

lr
lr train r2 score =  0.7511398012098071
lr test r2 score =  0.7359771482331381
lr train mean_squared_error =  -0.10476228008693259
lr test mean_squared_error =  -0.11053447526934486
------------------------------
Lasso
Lasso train r2 score =  0.0
Lasso test r2 score =  -0.01312694708708219
Lasso train mean_squared_error =  -0.4209058818109188
Lasso test mean_squared_error =  -0.42245064016356293
------------------------------
Ridge
Ridge train r2 score =  0.7511397917743586
Ridge test r2 score =  0.7359792899386175
Ridge train mean_squared_error =  -0.10476228405749917
Ridge test mean_squared_error =  -0.11053358948928206
------------------------------
SVC
SVC train r2 score =  0.9540451375517528
SVC test r2 score =  0.9024438663177436
SVC train mean_squared_error =  -0.01934121410781906
SVC test mean_squared_error =  -0.04068585926481661
------------------------------
Knn
Knn train r2 score =  0.8418569753756718
Knn test r2 score =  0.7379214962192625
Knn train mean_squared_error =

XGBoost parameter Hypertuning


In [22]:
PL=make_pipeline(transform,
                  StandardScaler(),

                  XGBRegressor())

XGB_params=[
    {
        'xgbregressor__learning_rate':[0.05,0.1,0.15,0.20,0.25,0.30],
        'xgbregressor__max_depth':[3,4,5,6,8,10,12,15],
        'xgbregressor__min_child_weight':[1,3,5,7],
        'xgbregressor__gamma':[0.0,0.1,0.2,0.3,0.4],
        'xgbregressor__colsample_bytree':[0.3,0.4,0.5,0.7]

    }
]

XGBRegressor_RS = RandomizedSearchCV(PL, XGB_params,n_iter=80, scoring='r2', cv=5,n_jobs=-1)
XGBRegressor_RS.fit(x,log_y)


RandomizedSearchCV(cv=5,
                   estimator=Pipeline(steps=[('columntransformer',
                                              ColumnTransformer(remainder='passthrough',
                                                                transformers=[('log',
                                                                               FunctionTransformer(func=<ufunc 'log1p'>),
                                                                               ['Kilometres']),
                                                                              ('Ohe',
                                                                               OneHotEncoder(drop='first',
                                                                                             sparse_output=False),
                                                                               ['UsedOrNew',
                                                                                'Transmission',
                                                                                'DriveType']),
                                                                              ('Be',
                                                                               BinaryEncoder(),
                                                                               ['FuelType',
                                                                                'ColourExtInt',
                                                                                'BodyType',
                                                                                'B...
                                                           num_parallel_tree=None,
                                                           random_state=None, ...))]),
                   n_iter=80, n_jobs=-1,
                   param_distributions=[{'xgbregressor__colsample_bytree': [0.3,
                                                                            0.4,
                                                                            0.5,
                                                                            0.7],
                                         'xgbregressor__gamma': [0.0, 0.1, 0.2,
                                                                 0.3, 0.4],
                                         'xgbregressor__learning_rate': [0.05,
                                                                         0.1,
                                                                         0.15,
                                                                         0.2,
                                                                         0.25,
                                                                         0.3],
                                         'xgbregressor__max_depth': [3, 4, 5, 6,
                                                                     8, 10, 12,
                                                                     15],
                                         'xgbregressor__min_child_weight': [1,
                                                                            3,
                                                                            5,
                                                                            7]}],
                   scoring='r2')

In [23]:
print("Best value for lambda : ",XGBRegressor_RS.best_params_)
print("Best score for cost function: ", XGBRegressor_RS.best_score_)

Best value for lambda :  {'xgbregressor__min_child_weight': 1, 'xgbregressor__max_depth': 10, 'xgbregressor__learning_rate': 0.15, 'xgbregressor__gamma': 0.0, 'xgbregressor__colsample_bytree': 0.7}
Best score for cost function:  0.9153683440678062


**Some result from RandomGridSearch for XGBoost**


Best value for lambda :  {'xgbregressor__min_child_weight': 7, 'xgbregressor__max_depth': 8, 'xgbregressor__learning_rate': 0.2, 'xgbregressor__gamma': 0.0, 'xgbregressor__colsample_bytree': 0.7}
Best score for cost function:  0.9150116958512857

Best value for lambda :  {'xgbregressor__min_child_weight': 3, 'xgbregressor__max_depth': 8, 'xgbregressor__learning_rate': 0.1, 'xgbregressor__gamma': 0.0, 'xgbregressor__colsample_bytree': 0.5}
Best score for cost function:  0.9165242053381277

Best value for lambda :  {'xgbregressor__min_child_weight': 1, 'xgbregressor__max_depth': 8, 'xgbregressor__learning_rate': 0.15, 'xgbregressor__gamma': 0.0, 'xgbregressor__colsample_bytree': 0.4}
Best score for cost function:  0.9146811705915328

In [ ]:
PL=make_pipeline(transform,
                  MinMaxScaler(),


                  XGBRegressor(min_child_weight= 1, max_depth= 8, learning_rate= 0.15, gamma= 0.0, colsample_bytree= 0.4))

cv=cross_validate(estimator=PL,X=x,y=log_y,cv=10,return_train_score=True,scoring=['r2','neg_mean_absolute_error'])
print('train r2 score = ',cv['train_r2'].mean())
print('test r2 score = ',cv['test_r2'].mean())
print('train mean_squared_error = ',cv['train_neg_mean_absolute_error'].mean())
print('test mean_squared_error = ',cv['test_neg_mean_absolute_error'].mean())


train r2 score =  0.9764584394081538
test r2 score =  0.9182238007545372
train mean_squared_error =  -0.07483486043311008
test mean_squared_error =  -0.12620708962312913


In [ ]:
PL=make_pipeline(transform,
                  StandardScaler(),


                  XGBRegressor(min_child_weight= 5, max_depth= 15, learning_rate= 0.05, gamma= 0.0, colsample_bytree= 0.7))

cv=cross_validate(estimator=PL,X=x,y=log_y,cv=10,return_train_score=True,scoring=['r2','neg_mean_absolute_error'])
print('train r2 score = ',cv['train_r2'].mean())
print('test r2 score = ',cv['test_r2'].mean())
print('train mean_squared_error = ',cv['train_neg_mean_absolute_error'].mean())
print('test mean_squared_error = ',cv['test_neg_mean_absolute_error'].mean())


train r2 score =  0.9920298989238872
test r2 score =  0.9156045900173755
train mean_squared_error =  -0.04025486181737707
test mean_squared_error =  -0.12618965954015687


**RandomForest parameter Hypertuning**

In [24]:
PL_RT=make_pipeline(transform,
                  StandardScaler(),

                  RandomForestRegressor())
RT_param = {

 'randomforestregressor__max_depth': [10, 20, 30, 40, 50, 60, 70, 80, 90, 100, None],
 'randomforestregressor__max_features': ['auto', 'sqrt'],
 'randomforestregressor__min_samples_leaf': [1, 2, 4],
 'randomforestregressor__min_samples_split': [2, 5, 10],
 'randomforestregressor__n_estimators': [200, 400, 600, 800, 1000, 1200, 1400, 1600, 1800, 2000]}
gs_RT = RandomizedSearchCV(PL_RT,RT_param,n_iter=20, cv=5,n_jobs=-1,)
gs_RT.fit(x,log_y)

RandomizedSearchCV(cv=5,
                   estimator=Pipeline(steps=[('columntransformer',
                                              ColumnTransformer(remainder='passthrough',
                                                                transformers=[('log',
                                                                               FunctionTransformer(func=<ufunc 'log1p'>),
                                                                               ['Kilometres']),
                                                                              ('Ohe',
                                                                               OneHotEncoder(drop='first',
                                                                                             sparse_output=False),
                                                                               ['UsedOrNew',
                                                                                'Transmission',
                                                                                'DriveType']),
                                                                              ('Be',
                                                                               BinaryEncoder(),
                                                                               ['FuelType',
                                                                                'ColourExtInt',
                                                                                'BodyType',
                                                                                'B...
                   n_iter=20, n_jobs=-1,
                   param_distributions={'randomforestregressor__max_depth': [10,
                                                                             20,
                                                                             30,
                                                                             40,
                                                                             50,
                                                                             60,
                                                                             70,
                                                                             80,
                                                                             90,
                                                                             100,
                                                                             None],
                                        'randomforestregressor__max_features': ['auto',
                                                                                'sqrt'],
                                        'randomforestregressor__min_samples_leaf': [1,
                                                                                    2,
                                                                                    4],
                                        'randomforestregressor__min_samples_split': [2,
                                                                                     5,
                                                                                     10],
                                        'randomforestregressor__n_estimators': [200,
                                                                                400,
                                                                                600,
                                                                                800,
                                                                                1000,
                                                                                1200,
                                                                                1400,
                                                                                1600,
                                                   

In [39]:
PL_RT=make_pipeline(transform,
                  StandardScaler(),

                  RandomForestRegressor())
RT_param = {

 'randomforestregressor__max_depth': [ 70, 80, None],
 'randomforestregressor__max_features': [ 'sqrt'],
 'randomforestregressor__min_samples_leaf': [1],
 'randomforestregressor__min_samples_split': [2],
 'randomforestregressor__n_estimators': [ 600, 800,  1800]}
gs_RT = GridSearchCV(PL_RT,RT_param, cv=10,n_jobs=-1,)
gs_RT.fit(x,log_y)

GridSearchCV(cv=10,
             estimator=Pipeline(steps=[('columntransformer',
                                        ColumnTransformer(remainder='passthrough',
                                                          transformers=[('log',
                                                                         FunctionTransformer(func=<ufunc 'log1p'>),
                                                                         ['Kilometres']),
                                                                        ('Ohe',
                                                                         OneHotEncoder(drop='first',
                                                                                       sparse_output=False),
                                                                         ['UsedOrNew',
                                                                          'Transmission',
                                                                          'DriveType']),
                                                                        ('Be',
                                                                         BinaryEncoder(),
                                                                         ['FuelType',
                                                                          'ColourExtInt',
                                                                          'BodyType',
                                                                          'Brand '...
                                       ('standardscaler', StandardScaler()),
                                       ('randomforestregressor',
                                        RandomForestRegressor())]),
             n_jobs=-1,
             param_grid={'randomforestregressor__max_depth': [70, 80, None],
                         'randomforestregressor__max_features': ['sqrt'],
                         'randomforestregressor__min_samples_leaf': [1],
                         'randomforestregressor__min_samples_split': [2],
                         'randomforestregressor__n_estimators': [600, 800,
                                                                 1800]})

In [40]:
print("Best value for lambda : ",gs_RT.best_params_)
print("Best score for cost function: ", gs_RT.best_score_)

Best value for lambda :  {'randomforestregressor__max_depth': 70, 'randomforestregressor__max_features': 'sqrt', 'randomforestregressor__min_samples_leaf': 1, 'randomforestregressor__min_samples_split': 2, 'randomforestregressor__n_estimators': 800}
Best score for cost function:  0.9117808084812025


In [41]:
gs_RT.refit_time_

23.59433364868164

**Some result from RandomGridSearch for Random Forest**

Best value for lambda :  {'randomforestregressor__n_estimators': 1800, 'randomforestregressor__min_samples_split': 2, 'randomforestregressor__min_samples_leaf': 1, 'randomforestregressor__max_features': 'sqrt', 'randomforestregressor__max_depth': 80}
Best score for cost function:  0.9184628907912666


Best value for lambda :  {'randomforestregressor__max_depth': 70, 'randomforestregressor__max_features': 'sqrt', 'randomforestregressor__min_samples_leaf': 1, 'randomforestregressor__min_samples_split': 2, 'randomforestregressor__n_estimators': 800}
Best score for cost function:  0.9117808084812025


In [28]:
PL=make_pipeline(transform,
                StandardScaler(),


                  RandomForestRegressor(n_estimators= 800,min_samples_split= 2,min_samples_leaf= 1,max_features= 'sqrt',max_depth=None,bootstrap= False))

cv=cross_validate(estimator=PL,X=x,y=log_y,cv=10,return_train_score=True,scoring=['r2','neg_mean_absolute_error'])
print('train r2 score = ',cv['train_r2'].mean())
print('test r2 score = ',cv['test_r2'].mean())
print('train mean_squared_error = ',cv['train_neg_mean_absolute_error'].mean())
print('test mean_absolute_error = ',cv['test_neg_mean_absolute_error'].mean())


train r2 score =  0.9999599515806767
test r2 score =  0.9183368924101032
train mean_squared_error =  -0.0002776147170465029
test mean_absolute_error =  -0.12346974307218313


In [29]:
PL=make_pipeline(transform,
                StandardScaler(),


                  RandomForestRegressor(n_estimators= 600,min_samples_split= 2,min_samples_leaf= 1,max_features= 'sqrt',max_depth=None,bootstrap= False))

cv=cross_validate(estimator=PL,X=x,y=log_y,cv=10,return_train_score=True,scoring=['r2','neg_mean_absolute_error'])
print('train r2 score = ',cv['train_r2'].mean())
print('test r2 score = ',cv['test_r2'].mean())
print('train mean_squared_error = ',cv['train_neg_mean_absolute_error'].mean())
print('test mean_absolute_error = ',cv['test_neg_mean_absolute_error'].mean())


train r2 score =  0.9999599515813357
test r2 score =  0.918080095089991
train mean_squared_error =  -0.0002776119009775146
test mean_absolute_error =  -0.12369856445465972


**Bagging parameter Hypertuning**

In [51]:
PL_BA=make_pipeline(transform,
                  StandardScaler(),

                  BaggingRegressor())

BA_params=[
    {
        'baggingregressor__bootstrap': [True,False],

'baggingregressor__max_features': [1.0,2.0,3.0],
'baggingregressor__max_samples': [1.0,1.5,2.0,3.0],
'baggingregressor__n_estimators': [10,50,100,1000]

    }
]

Bagging_Regressor = RandomizedSearchCV(PL_BA, BA_params,n_iter=70, scoring='r2', cv=10,n_jobs=-1)
Bagging_Regressor.fit(x,log_y)


RandomizedSearchCV(cv=10,
                   estimator=Pipeline(steps=[('columntransformer',
                                              ColumnTransformer(remainder='passthrough',
                                                                transformers=[('log',
                                                                               FunctionTransformer(func=<ufunc 'log1p'>),
                                                                               ['Kilometres']),
                                                                              ('Ohe',
                                                                               OneHotEncoder(drop='first',
                                                                                             sparse_output=False),
                                                                               ['UsedOrNew',
                                                                                'Transmission',
                                                                                'DriveType']),
                                                                              ('Be',
                                                                               BinaryEncoder(),
                                                                               ['FuelType',
                                                                                'ColourExtInt',
                                                                                'BodyType',
                                                                                '...'
                                                                                '& '
                                                                                'Model',
                                                                                'State'])])),
                                             ('standardscaler',
                                              StandardScaler()),
                                             ('baggingregressor',
                                              BaggingRegressor())]),
                   n_iter=70, n_jobs=-1,
                   param_distributions=[{'baggingregressor__bootstrap': [True,
                                                                         False],
                                         'baggingregressor__max_features': [1.0,
                                                                            2.0,
                                                                            3.0],
                                         'baggingregressor__max_samples': [1.0,
                                                                           1.5,
                                                                           2.0,
                                                                           3.0],
                                         'baggingregressor__n_estimators': [10,
                                                                            50,
                                                                            100,
                                                                            1000]}],
                   scoring='r2')

In [52]:
print("Best value for lambda : ",Bagging_Regressor.best_params_)
print("Best score for cost function: ", Bagging_Regressor.best_score_)

Best value for lambda :  {'baggingregressor__n_estimators': 100, 'baggingregressor__max_samples': 1.0, 'baggingregressor__max_features': 1.0, 'baggingregressor__bootstrap': True}
Best score for cost function:  0.9021078108883603


Some result from RandomGridSearch for Bagging

Best value for lambda :  {'baggingregressor__n_estimators': 1000, 'baggingregressor__max_samples': 1.0, 'baggingregressor__max_features': 1.0, 'baggingregressor__bootstrap': True}
Best score for cost function:  0.8315640606942826



In [ ]:
PL=make_pipeline(transform,
                StandardScaler(),

                BaggingRegressor(n_estimators= 1000,max_samples= 1.0,max_features= 1.0,bootstrap= True))

cv=cross_validate(estimator=PL,X=x,y=log_y,cv=10,return_train_score=True,scoring=['r2','neg_mean_absolute_error'])
print('train r2 score = ',cv['train_r2'].mean())
print('test r2 score = ',cv['test_r2'].mean())
print('train mean_squared_error = ',cv['train_neg_mean_absolute_error'].mean())
print('test mean_absolute_error = ',cv['test_neg_mean_absolute_error'].mean())

train r2 score =  0.9872103915986095
test r2 score =  0.903754689579573
train mean_squared_error =  -0.04960453552880141
test mean_absolute_error =  -0.13625463267603385


**Saving the model from Random forest**

In [42]:
import joblib

In [43]:
joblib.dump(gs_RT.best_estimator_ , 'Model.h5')

['Model.h5']

Save it in google drive

In [44]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive
